bls-signatures/src/signature.rs

```diff
@@ -66,58 +68,92 @@ impl SignatureProjective {
         for signature in signatures {
-            self.0 += &signature.try_as_projective()?.0;
+            self.0 += signature.try_as_projective()?.0;
         }
         Ok(())
     }
 
-    /// Aggregate a list of public keys
+    /// Aggregate a list of signatures
     #[allow(clippy::arithmetic_side_effects)]
-    pub fn aggregate<'a, S: 'a + AsSignatureProjective + ?Sized, I>(
-        signatures: I,
-    ) -> Result<SignatureProjective, BlsError>
-    where
-        I: IntoIterator<Item = &'a S>,
-    {
-        let mut iter = signatures.into_iter();
-        if let Some(first) = iter.next() {
+    pub fn aggregate<S: AsSignatureProjective + ?Sized>(
+        signatures: &[&S],
+    ) -> Result<SignatureProjective, BlsError> {
+        if let Some((first, rest)) = signatures.split_first() {
             let mut aggregate = first.try_as_projective()?;
-            aggregate.aggregate_with(iter)?;
+            aggregate.aggregate_with(rest)?;
             Ok(aggregate)
         } else {
             Err(BlsError::EmptyAggregation)
         }
     }
 
     /// Verify a list of signatures against a message and a list of public keys
-    pub fn aggregate_verify<
-        'a,
-        P: 'a + AsPubkeyProjective + ?Sized,
-        S: 'a + AsSignatureProjective + ?Sized,
-        I,
-        J,
-    >(
-        public_keys: I,
-        signatures: J,
+    pub fn aggregate_verify<P: AsPubkeyProjective + ?Sized, S: AsSignatureProjective + ?Sized>(
+        public_keys: &[&P],
+        signatures: &[&S],
         message: &[u8],
-    ) -> Result<bool, BlsError>
-    where
-        I: IntoIterator<Item = &'a P>,
-        J: IntoIterator<Item = &'a S>,
-    {
+    ) -> Result<bool, BlsError> {
         let aggregate_pubkey = PubkeyProjective::aggregate(public_keys)?;
         let aggregate_signature = SignatureProjective::aggregate(signatures)?;
 
         Ok(aggregate_pubkey._verify_signature(&aggregate_signature, message))
     }
+
+    /// Aggregate a list of signatures into an existing aggregate
+    #[allow(clippy::arithmetic_side_effects)]
+    #[cfg(feature = "parallel")]
+    pub fn par_aggregate_with<S: AsSignatureProjective + Sync>(
+        &mut self,
+        signatures: &[&S],
+    ) -> Result<(), BlsError> {
+        let aggregate = SignatureProjective::par_aggregate(signatures)?;
+        self.0 += &aggregate.0;
+        Ok(())
+    }
+
+    /// Aggregate a list of signatures
+    #[allow(clippy::arithmetic_side_effects)]
+    #[cfg(feature = "parallel")]
+    pub fn par_aggregate<S: AsSignatureProjective + Sync>(
+        signatures: &[&S],
+    ) -> Result<SignatureProjective, BlsError> {
+        if signatures.is_empty() {
+            return Err(BlsError::EmptyAggregation);
+        }
+        signatures
+            .into_par_iter()
+            .map(|sig| (*sig).try_as_projective())
+            .reduce(
+                || Ok(SignatureProjective::identity()),
+                |a, b| {
+                    let mut a = a?;
```

Comment:
I guess if `let mut a = a?;` fails we have a big trouble.... But I'm fine with it written this way.

Reply:
Yeah assuming that all the signatures have been pre-verified and valid, this should never fail. We have to unwrap the logic here somehow and I think this is probably the cleanest way to write it.